File: VB/ControlTemplateGallerySample/ControlTemplateGallery/TemplateStorage.vb
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Data
Imports DevExpress.XtraEditors
Imports System.Windows.Forms
Imports System.Drawing

Namespace ControlTemplateGallerySample
	Public Interface IControlTemplateStorage
		Function GetCategoryNames() As String()
		Function GetTemplateNamesForCategory(ByVal categoryName As String) As String()
		Function Restore(ByVal categoryName As String, ByVal templateName As String) As XRControl()
		Sub Store(ByVal categoryName As String, ByVal templateName As String, ByVal controls() As XRControl)
		Sub DeleteTemplate(ByVal categoryName As String, ByVal templateName As String)
		Sub DeleteCategory(ByVal categoryName As String)
	End Interface

	Friend Class ControlTemplateStorageBase
		Protected Shared Function DeserializeControls(ByVal bytes() As Byte) As XRControl()
			If bytes IsNot Nothing AndAlso bytes.Length > 0 Then
				Dim tempReport As New XtraReport()
				Using stream As New MemoryStream(bytes)
					tempReport.LoadLayoutFromXml(stream)
				End Using
				Return tempReport.Bands(BandKind.Detail).Controls.Cast(Of XRControl)().ToArray()
			End If
			Return New XRControl(){}
		End Function
		Protected Shared Function SerializeControls(ByVal controls() As XRControl) As Byte()
			Dim tempReport As New XtraReport()
			Dim tempBand As New DetailBand()
			tempReport.Bands.Add(tempBand)

			Using cloneReport As New XtraReport()
				Using stream As New MemoryStream()
					controls(0).RootReport.SaveLayoutToXml(stream)
					stream.Position = 0

					cloneReport.LoadLayoutFromXml(stream)
				End Using
				For Each control As XRControl In controls
					Dim clone As XRControl = cloneReport.FindControl(control.Name, True)
					If clone IsNot Nothing Then
						tempBand.Controls.Add(clone)
					End If
				Next control
			End Using

			Dim topMost As Single = Single.MaxValue
			Dim leftMost As Single = Single.MaxValue

			For Each control As XRControl In tempBand.Controls
				topMost = Math.Min(topMost, control.TopF)
				leftMost = Math.Min(leftMost, control.LeftF)
			Next control

			For Each control As XRControl In tempBand.Controls
				control.LocationF = New PointF(control.LeftF - leftMost, control.TopF - topMost)
			Next control

			Dim layoutBytes() As Byte = Nothing

			Using stream As New MemoryStream()
				tempReport.SaveLayoutToXml(stream)
				layoutBytes = stream.ToArray()
			End Using

			tempReport.Dispose()
			Return layoutBytes
		End Function
	End Class

	' Persistent storage
	Friend Class PersistentControlTemplateStorage
		Inherits ControlTemplateStorageBase
		Implements IControlTemplateStorage

		Private ds As DataSet

		Public Sub New(ByVal path As String)
			ds = New DataSet()
			If File.Exists(path) Then
				ds.ReadXml(path)
			Else
				Dim categories As New DataTable("Categories")
				ds.Tables.Add(categories)

				categories.Columns.Add("CategoryID", GetType(Integer)).AutoIncrement = True
				categories.Columns.Add("CategoryName", GetType(String))

				Dim templates As New DataTable("Templates")
				ds.Tables.Add(templates)

				templates.Columns.Add("CategoryID", GetType(Integer))
				templates.Columns.Add("TemplateName", GetType(String))
				templates.Columns.Add("Layout", GetType(Byte()))

				ds.Relations.Add("CategoryTemplates", categories.Columns("CategoryID"), templates.Columns("CategoryID"))

				ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema)
			End If
		End Sub

	   Public Sub DeleteTemplate(ByVal categoryName As String, ByVal templateName As String) Implements IControlTemplateStorage.DeleteTemplate
			Dim row As DataRow = ds.Tables("Categories").AsEnumerable().Where(Function(x) x.Field(Of String)("CategoryName") = categoryName).FirstOrDefault()

			Dim catId As Integer = row.Field(Of Integer)("CategoryID")
			Dim templateRow As DataRow = ds.Tables("Templates").AsEnumerable().Where(Function(x) x.Field(Of Integer)("CategoryID") = catId AndAlso x.Field(Of String)("TemplateName") = templateName).FirstOrDefault()
			ds.Tables("Templates").Rows.Remove(templateRow)

			ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema)
	   End Sub

	   Public Sub DeleteCategory(ByVal categoryName As String) Implements IControlTemplateStorage.DeleteCategory
			Dim row As DataRow = ds.Tables("Categories").AsEnumerable().Where(Function(x) x.Field(Of String)("CategoryName") = categoryName).FirstOrDefault()

			Dim templateRows() As DataRow = row.GetChildRows("CategoryTempaltes")
			For i As Integer = 0 To templateRows.Length - 1
				ds.Tables("Templates").Rows.Remove(templateRows(i))
			Next i

			ds.Tables("Categories").Rows.Remove(row)

			ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema)
	   End Sub

	   Public Function GetCategoryNames() As String() Implements IControlTemplateStorage.GetCategoryNames
			Return ds.Tables("Categories").AsEnumerable().Select(Function(r) r.Field(Of String)("CategoryName")).ToArray()
	   End Function

	   Public Function GetTemplateNamesForCategory(ByVal categoryName As String) As String() Implements IControlTemplateStorage.GetTemplateNamesForCategory
			Dim row As DataRow = ds.Tables("Categories").AsEnumerable().Where(Function(x) x.Field(Of String)("CategoryName") = categoryName).FirstOrDefault()
			If row IsNot Nothing Then
				Return row.GetChildRows("CategoryTemplates").Select(Function(r) r.Field(Of String)("TemplateName")).ToArray()
			End If

			Return Nothing
	   End Function

	   Public Function Restore(ByVal categoryName As String, ByVal templateName As String) As XRControl() Implements IControlTemplateStorage.Restore
			Dim row As DataRow = ds.Tables("Categories").AsEnumerable().Where(Function(x) x.Field(Of String)("CategoryName") = categoryName).FirstOrDefault()
			If row IsNot Nothing Then
				Dim childRow As DataRow = row.GetChildRows("CategoryTemplates").Where(Function(x) x.Field(Of String)("TemplateName") = templateName).FirstOrDefault()
				Dim layout() As Byte = If(childRow IsNot Nothing, childRow.Field(Of Byte())("Layout"), New Byte(){})
				Return DeserializeControls(layout)
			End If
			Return Nothing
	   End Function

	   Public Sub Store(ByVal categoryName As String, ByVal templateName As String, ByVal controls() As XRControl) Implements IControlTemplateStorage.Store
		   Dim templateLayout() As Byte = SerializeControls(controls)

			Dim row As DataRow = ds.Tables("Categories").AsEnumerable().Where(Function(x) x.Field(Of String)("CategoryName") = categoryName).FirstOrDefault()
			If row Is Nothing Then
				row = ds.Tables("Categories").NewRow()
				row("CategoryName") = categoryName
				ds.Tables("Categories").Rows.Add(row)
			End If

			Dim catId As Integer = row.Field(Of Integer)("CategoryID")
			Dim templateRow As DataRow = ds.Tables("Templates").AsEnumerable().Where(Function(x) x.Field(Of Integer)("CategoryID") = catId AndAlso x.Field(Of String)("TemplateName") = templateName).FirstOrDefault()
			If templateRow Is Nothing Then
				templateRow = ds.Tables("Templates").NewRow()
				templateRow("CategoryID") = catId
				templateRow("TemplateName") = templateName
				ds.Tables("Templates").Rows.Add(templateRow)
			End If
			templateRow("Layout") = templateLayout

			ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema)
	   End Sub
	End Class

	' In-memory storage
	Friend Class ControlTemplateStorage
		Inherits ControlTemplateStorageBase
		Implements IControlTemplateStorage

		Private Class TemplateLayoutItem
			Public Property Category() As String
			Public Property Name() As String
			Public Property LayoutBytes() As Byte()
		End Class

		Private templates As List(Of TemplateLayoutItem)

		Public Sub New()
			templates = New List(Of TemplateLayoutItem)()
		End Sub

		Public Sub DeleteTemplate(ByVal categoryName As String, ByVal templateName As String) Implements IControlTemplateStorage.DeleteTemplate
			Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
			templates.Remove(item)
		End Sub

		Public Sub DeleteCategory(ByVal categoryName As String) Implements IControlTemplateStorage.DeleteCategory
			Dim items As IEnumerable(Of TemplateLayoutItem) = templates.Where(Function(x) x.Category = categoryName)

			For Each item As TemplateLayoutItem In items
				templates.Remove(item)
			Next item
		End Sub

		Public Function GetCategoryNames() As String() Implements IControlTemplateStorage.GetCategoryNames
			Return templates.Select(Function(x) x.Category).Distinct().ToArray()
		End Function

		Public Function GetTemplateNamesForCategory(ByVal categoryName As String) As String() Implements IControlTemplateStorage.GetTemplateNamesForCategory
			Return templates.Where(Function(x) x.Category = categoryName).Select(Function(y) y.Name).ToArray()
		End Function

		Public Function Restore(ByVal categoryName As String, ByVal templateName As String) As XRControl() Implements IControlTemplateStorage.Restore
			Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
			Dim layout() As Byte = If(item IsNot Nothing, item.LayoutBytes, New Byte(){})
			Return DeserializeControls(layout)
		End Function

		Public Sub Store(ByVal categoryName As String, ByVal templateName As String, ByVal controls() As XRControl) Implements IControlTemplateStorage.Store
			Dim templateLayout() As Byte = SerializeControls(controls)
			Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
			If item Is Nothing Then
				templates.Add(New TemplateLayoutItem() With {.Name = templateName, .Category = categoryName, .LayoutBytes = templateLayout})
			End If
		End Sub
	End Class
End Namespace

File: CS/ControlTemplateGallerySample/ControlTemplateGallery/TemplateStorage.cs
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Data;
using DevExpress.XtraEditors;
using System.Windows.Forms;
using System.Drawing;

namespace ControlTemplateGallerySample
{
    public interface IControlTemplateStorage
    {
        string[] GetCategoryNames();
        string[] GetTemplateNamesForCategory(string categoryName);
        XRControl[] Restore(string categoryName, string templateName);
        void Store(string categoryName, string templateName, XRControl[] controls);
        void DeleteTemplate(string categoryName, string templateName);
        void DeleteCategory(string categoryName);
    }

    class ControlTemplateStorageBase {
        protected static XRControl[] DeserializeControls(byte[] bytes) {
            if(bytes != null && bytes.Length > 0) {
                XtraReport tempReport = new XtraReport();
                using(MemoryStream stream = new MemoryStream(bytes)) {
                    tempReport.LoadLayoutFromXml(stream);
                }
                return tempReport.Bands[BandKind.Detail].Controls.Cast<XRControl>().ToArray();
            }
            return new XRControl[0];
        }
        protected static byte[] SerializeControls(XRControl[] controls) {
            XtraReport tempReport = new XtraReport();
            DetailBand tempBand = new DetailBand();
            tempReport.Bands.Add(tempBand);

            using(XtraReport cloneReport = new XtraReport()) {
                using(MemoryStream stream = new MemoryStream()) {
                    controls[0].RootReport.SaveLayoutToXml(stream);
                    stream.Position = 0;

                    cloneReport.LoadLayoutFromXml(stream);
                }
                foreach(XRControl control in controls) {
                    XRControl clone = cloneReport.FindControl(control.Name, true);
                    if(clone != null) tempBand.Controls.Add(clone);
                }
            }

            float topMost = float.MaxValue;
            float leftMost = float.MaxValue;

            foreach(XRControl control in tempBand.Controls) {
                topMost = Math.Min(topMost, control.TopF);
                leftMost = Math.Min(leftMost, control.LeftF);
            }

            foreach(XRControl control in tempBand.Controls) {
                control.LocationF = new PointF(control.LeftF - leftMost, control.TopF - topMost);
            }

            byte[] layoutBytes = null;

            using(MemoryStream stream = new MemoryStream()) {
                tempReport.SaveLayoutToXml(stream);
                layoutBytes = stream.ToArray();
            }

            tempReport.Dispose();
            return layoutBytes;
        }
    }

    // Persistent storage
    class PersistentControlTemplateStorage : ControlTemplateStorageBase, IControlTemplateStorage {
        DataSet ds;

        public PersistentControlTemplateStorage(string path) {
            ds = new DataSet();
            if(File.Exists(path))
                ds.ReadXml(path);
            else {
                DataTable categories = new DataTable("Categories");
                ds.Tables.Add(categories);

                categories.Columns.Add("CategoryID", typeof(int)).AutoIncrement = true;
                categories.Columns.Add("CategoryName", typeof(string));

                DataTable templates = new DataTable("Templates");
                ds.Tables.Add(templates);

                templates.Columns.Add("CategoryID", typeof(int));
                templates.Columns.Add("TemplateName", typeof(string));
                templates.Columns.Add("Layout", typeof(byte[]));

                ds.Relations.Add("CategoryTemplates", categories.Columns["CategoryID"], templates.Columns["CategoryID"]);

                ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema);
            }
        }

       public void DeleteTemplate(string categoryName, string templateName) {
            DataRow row = ds.Tables["Categories"].AsEnumerable().Where(x => x.Field<string>("CategoryName") == categoryName).FirstOrDefault();

            int catId = row.Field<int>("CategoryID");
            DataRow templateRow = ds.Tables["Templates"].AsEnumerable().Where(x => x.Field<int>("CategoryID") == catId && x.Field<string>("TemplateName") == templateName).FirstOrDefault();
            ds.Tables["Templates"].Rows.Remove(templateRow);

            ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema);
        }

       public void DeleteCategory(string categoryName) {
            DataRow row = ds.Tables["Categories"].AsEnumerable().Where(x => x.Field<string>("CategoryName") == categoryName).FirstOrDefault();

            DataRow[] templateRows = row.GetChildRows("CategoryTempaltes");
            for(int i = 0; i < templateRows.Length; i++)
                ds.Tables["Templates"].Rows.Remove(templateRows[i]);

            ds.Tables["Categories"].Rows.Remove(row);

            ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema);
        }

       public string[] GetCategoryNames() {
            return ds.Tables["Categories"].AsEnumerable().Select(r => r.Field<string>("CategoryName")).ToArray();
        }

       public string[] GetTemplateNamesForCategory(string categoryName) {
            DataRow row = ds.Tables["Categories"].AsEnumerable().Where(x => x.Field<string>("CategoryName") == categoryName).FirstOrDefault();
            if(row != null)
                return row.GetChildRows("CategoryTemplates").Select(r => r.Field<string>("TemplateName")).ToArray();

            return null;
        }

       public XRControl[] Restore(string categoryName, string templateName) {
            DataRow row = ds.Tables["Categories"].AsEnumerable().Where(x => x.Field<string>("CategoryName") == categoryName).FirstOrDefault();
            if(row != null) {
                DataRow childRow = row.GetChildRows("CategoryTemplates").Where(x => x.Field<string>("TemplateName") == templateName).FirstOrDefault();
                byte[] layout = childRow != null ? childRow.Field<byte[]>("Layout") : new byte[0];
                return DeserializeControls(layout);
            }
            return null;
        }

       public void Store(string categoryName, string templateName, XRControl[] controls) {
           byte[] templateLayout = SerializeControls(controls);

            DataRow row = ds.Tables["Categories"].AsEnumerable().Where(x => x.Field<string>("CategoryName") == categoryName).FirstOrDefault();
            if(row == null) {
                row = ds.Tables["Categories"].NewRow();
                row["CategoryName"] = categoryName;
                ds.Tables["Categories"].Rows.Add(row);
            }

            int catId = row.Field<int>("CategoryID");
            DataRow templateRow = ds.Tables["Templates"].AsEnumerable().Where(x => x.Field<int>("CategoryID") == catId && x.Field<string>("TemplateName") == templateName).FirstOrDefault();
            if(templateRow == null) {
                templateRow = ds.Tables["Templates"].NewRow();
                templateRow["CategoryID"] = catId;
                templateRow["TemplateName"] = templateName;
                ds.Tables["Templates"].Rows.Add(templateRow);
            }
            templateRow["Layout"] = templateLayout;

            ds.WriteXml("SampleStorage.xml", XmlWriteMode.WriteSchema);
        }
    }

    // In-memory storage
    class ControlTemplateStorage : ControlTemplateStorageBase, IControlTemplateStorage {
        class TemplateLayoutItem {
            public string Category { get; set; }
            public string Name { get; set; }
            public byte[] LayoutBytes { get; set; }
        }

        List<TemplateLayoutItem> templates;

        public ControlTemplateStorage() {
            templates = new List<TemplateLayoutItem>();
        }

        public void DeleteTemplate(string categoryName, string templateName) {
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            templates.Remove(item);
        }

        public void DeleteCategory(string categoryName) {
            IEnumerable<TemplateLayoutItem> items = templates.Where(x => x.Category == categoryName);

            foreach(TemplateLayoutItem item in items)
                templates.Remove(item);
        }

        public string[] GetCategoryNames() {
            return templates.Select(x => x.Category).Distinct().ToArray();
        }

        public string[] GetTemplateNamesForCategory(string categoryName) {
            return templates.Where(x => x.Category == categoryName).Select(y => y.Name).ToArray();
        }

        public XRControl[] Restore(string categoryName, string templateName) {
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            byte[] layout = item != null ? item.LayoutBytes : new byte[0];
            return DeserializeControls(layout);
        }

        public void Store(string categoryName, string templateName, XRControl[] controls) {
            byte[] templateLayout = SerializeControls(controls);
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            if(item == null)
                templates.Add(new TemplateLayoutItem() { Name = templateName, Category = categoryName, LayoutBytes = templateLayout });
        }
    }
}
